File: services/payment-service/src/services/account.service.ts
import { config } from '../config';
import { pool } from '../db/pool';
import { withTransaction } from '../db/transaction';
import type { AccountLimits } from '../domain/limits';
import { HttpError, notFound } from '../lib/http-error';
import { toAccountDto, type AccountDto } from '../models/account.model';
import * as accounts from '../repositories/account.repository';
import * as ledger from '../repositories/ledger.repository';
import * as outbox from '../repositories/outbox.repository';

/**
 * Opens a wallet.
 *
 * An opening balance is not money appearing from nowhere: it is issued by the
 * funding account, which goes negative by exactly this much. That is what
 * makes the ledger complete - every cent has a provenance, and the balances of
 * all accounts together still sum to zero, which is the invariant the
 * reconciliation control leans on.
 */
export async function createAccount(
  name: string,
  initialBalanceCents: number,
): Promise<AccountDto> {
  const row = await withTransaction(async (client) => {
    const account = await accounts.insert(client, name, initialBalanceCents);

    if (initialBalanceCents > 0) {
      const locked = await accounts.lockMany(client, [config.systemAccounts.fundingId]);
      const funding = locked.get(config.systemAccounts.fundingId);
      if (!funding) throw new HttpError(500, 'FUNDING_ACCOUNT_MISSING');

      await accounts.updateBalance(
        client,
        funding.id,
        funding.balanceCents - initialBalanceCents,
      );
      await ledger.postJournal(client, null, 'FUNDING', [
        { accountId: funding.id, direction: 'DEBIT', amountCents: initialBalanceCents },
        { accountId: account.id, direction: 'CREDIT', amountCents: initialBalanceCents },
      ]);
    }

    await outbox.enqueue(client, 'account.created', {
      accountId: account.id,
      name: account.name,
      balanceCents: account.balance_cents,
      occurredAt: account.created_at.toISOString(),
    });
    return account;
  });

  return toAccountDto(row);
}

/**
 * The system accounts are plumbing, not people - hidden from the wallet's
 * friends list, visible to the developer dashboard.
 */
export async function listAccounts(includeSystem: boolean): Promise<AccountDto[]> {
  const rows = await accounts.findAll(pool, includeSystem);
  return rows.map(toAccountDto);
}

export async function getAccount(id: string): Promise<AccountDto> {
  const row = await accounts.findById(pool, id);
  if (!row) throw notFound('ACCOUNT_NOT_FOUND');
  return toAccountDto(row);
}

export interface AccountLimitsView {
  accountId: string;
  limits: AccountLimits;
  usage: {
    todayCents: number;
    recentCount: number;
    windowSeconds: number;
  };
  /** Headroom left against the daily cap, so a client can warn before sending. */
  remainingTodayCents: number;
}

/**
 * An account's spending controls and how much of them it has used.
 *
 * Read outside a transaction and without a lock: this is for showing someone
 * their headroom, not for deciding anything. The number that matters is
 * recomputed under the sender's row lock at authorise time, which is the only
 * place it can be trusted.
 */
export async function getLimits(id: string): Promise<AccountLimitsView> {
  const limits = await accounts.findLimits(pool, id);
  if (!limits) throw notFound('ACCOUNT_NOT_FOUND');

  const { velocityWindowSeconds } = config.controls;
  const usage = await accounts.spendSoFar(pool, id, velocityWindowSeconds);

  return {
    accountId: id,
    limits,
    usage: { ...usage, windowSeconds: velocityWindowSeconds },
    remainingTodayCents: Math.max(0, limits.dailyLimitCents - usage.todayCents),
  };
}

/** System accounts are plumbing and have no spending controls to set. */
export async function setLimits(
  id: string,
  limits: AccountLimits,
): Promise<AccountLimitsView> {
  const updated = await withTransaction((client) =>
    accounts.updateLimits(client, id, limits),
  );
  if (!updated) throw notFound('ACCOUNT_NOT_FOUND');
  return getLimits(id);
}

File: services/payment-service/src/main/java/com/ledgerflow/payment/services/AccountService.java
package com.ledgerflow.payment.services;

import com.ledgerflow.payment.config.Config;
import com.ledgerflow.payment.db.Tx;
import com.ledgerflow.payment.domain.Direction;
import com.ledgerflow.payment.domain.Leg;
import com.ledgerflow.payment.domain.Limits.AccountLimits;
import com.ledgerflow.payment.domain.Limits.SpendSoFar;
import com.ledgerflow.payment.domain.Payments.Account;
import com.ledgerflow.payment.domain.Payments.LedgerEntry;
import com.ledgerflow.payment.lib.HttpError;
import com.ledgerflow.payment.lib.Iso;
import com.ledgerflow.payment.models.AccountModel;
import com.ledgerflow.payment.models.AccountModel.AccountDto;
import com.ledgerflow.payment.models.AccountModel.AccountRow;
import com.ledgerflow.payment.repositories.AccountRepository;
import com.ledgerflow.payment.repositories.LedgerRepository;
import com.ledgerflow.payment.repositories.OutboxRepository;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class AccountService {

  /**
   * @param remainingTodayCents headroom left against the daily cap, so a client
   *     can warn before sending.
   */
  public record AccountLimitsView(
      String accountId, AccountLimits limits, Usage usage, long remainingTodayCents) {

    public record Usage(long todayCents, int recentCount, int windowSeconds) {}
  }

  private final Tx tx;
  private final AccountRepository accounts;
  private final LedgerRepository ledger;
  private final OutboxRepository outbox;

  public AccountService(
      Tx tx, AccountRepository accounts, LedgerRepository ledger, OutboxRepository outbox) {
    this.tx = tx;
    this.accounts = accounts;
    this.ledger = ledger;
    this.outbox = outbox;
  }

  /**
   * Opens a wallet.
   *
   * An opening balance is not money appearing from nowhere: it is issued by the
   * funding account, which goes negative by exactly this much. That is what
   * makes the ledger complete - every cent has a provenance, and the balances of
   * all accounts together still sum to zero, which is the invariant the
   * reconciliation control leans on.
   */
  public AccountDto createAccount(String name, long initialBalanceCents) {
    AccountRow row =
        tx.inTransaction(
            () -> {
              AccountRow account = accounts.insert(name, initialBalanceCents);

              if (initialBalanceCents > 0) {
                Map<String, Account> locked =
                    accounts.lockMany(List.of(Config.SystemAccounts.FUNDING_ID));
                Account funding = locked.get(Config.SystemAccounts.FUNDING_ID);
                if (funding == null) throw new HttpError(500, "FUNDING_ACCOUNT_MISSING");

                accounts.updateBalance(
                    funding.id(), funding.balanceCents() - initialBalanceCents);
                ledger.postJournal(
                    null,
                    Leg.FUNDING,
                    List.of(
                        new LedgerEntry(funding.id(), Direction.DEBIT, initialBalanceCents),
                        new LedgerEntry(account.id(), Direction.CREDIT, initialBalanceCents)));
              }

              Map<String, Object> event = new LinkedHashMap<>();
              event.put("accountId", account.id());
              event.put("name", account.name());
              event.put("balanceCents", account.balanceCents());
              event.put("occurredAt", Iso.format(account.createdAt()));
              outbox.enqueue("account.created", event);
              return account;
            });

    return AccountModel.toAccountDto(row);
  }

  /**
   * The system accounts are plumbing, not people - hidden from the wallet's
   * friends list, visible to the developer dashboard.
   */
  public List<AccountDto> listAccounts(boolean includeSystem) {
    List<AccountDto> dtos = new ArrayList<>();
    for (AccountRow row : accounts.findAll(includeSystem)) {
      dtos.add(AccountModel.toAccountDto(row));
    }
    return dtos;
  }

  public AccountDto getAccount(String id) {
    AccountRow row = accounts.findById(id);
    if (row == null) throw HttpError.notFound("ACCOUNT_NOT_FOUND");
    return AccountModel.toAccountDto(row);
  }

  /**
   * An account's spending controls and how much of them it has used.
   *
   * Read outside a transaction and without a lock: this is for showing someone
   * their headroom, not for deciding anything. The number that matters is
   * recomputed under the sender's row lock at authorise time, which is the only
   * place it can be trusted.
   */
  public AccountLimitsView getLimits(String id) {
    AccountLimits limits = accounts.findLimits(id);
    if (limits == null) throw HttpError.notFound("ACCOUNT_NOT_FOUND");

    int windowSeconds = Config.Controls.VELOCITY_WINDOW_SECONDS;
    SpendSoFar usage = accounts.spendSoFar(id, windowSeconds);

    return new AccountLimitsView(
        id,
        limits,
        new AccountLimitsView.Usage(usage.todayCents(), usage.recentCount(), windowSeconds),
        Math.max(0, limits.dailyLimitCents() - usage.todayCents()));
  }

  /** System accounts are plumbing and have no spending controls to set. */
  public AccountLimitsView setLimits(String id, AccountLimits limits) {
    AccountRow updated = tx.inTransaction(() -> accounts.updateLimits(id, limits));
    if (updated == null) throw HttpError.notFound("ACCOUNT_NOT_FOUND");
    return getLimits(id);
  }
}
